### PR TITLE
Enhancements for Zyxel OS 

### PR DESCRIPTION
**Pull request by Jimmy-Cl**
Please give a short description what your pull request is for

Enhancements for Zyxel OS (ZYNOS) -
- Human Readable port numbers
- Added Dbm levels to ports
- Added Stacking state sensor

I had an old pull request #13750 however I decided to close it while I made major changes to my port script.

image

image

image

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [x] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by Jimmy-Cl** [comment]
Passed tests on my repo, Please can someone approve the workflows here for testing.

File: includes/definitions/discovery/zynos.yaml [hunk 1]
```diff
@@ -66,35 +115,62 @@ modules:
...
-                    num_oid: '.1.3.6.1.4.1.890.1.5.8.47.117.2.1.7.{{ $index }}'
-                    descr: 'SFP ifIndex {{ $transceiverDdmInfoEntryPort }} {{ $transceiverDdmInfoEntryDescription }}'
-                    index: 'transceiverDdmInfoEntryCurrent.{{ $index }}'
-                    high_limit: transceiverDdmInfoEntryAlarmMax
-                    low_limit: transceiverDdmInfoEntryAlarmMin
-                    warn_limit: transceiverDdmInfoEntryWarnMax
-                    low_warn_limit: transceiverDdmInfoEntryWarnMin
-                    divisor: 100
-                    entPhysicalIndex: transceiverDdmInfoEntryPort
-                    entPhysicalIndex_measured: ports
```
Comment (Jellyfrog) [inline includes/definitions/discovery/zynos.yaml:82]: are you sure nothing is using this?

File: includes/definitions/discovery/zynos.yaml [hunk 2]
```diff
@@ -104,25 +180,7 @@ modules:
                     value: voltageCurValue
                     snmp_flags: ['-OteQUs', '-Pu'] # workaround for underscores in mib
                     num_oid: '.1.3.6.1.4.1.890.1.5.8.47.8.3.1.2.{{ $index }}'
+                    divisor: 1000
                     descr: '{{ $voltageNominalValue }}mV ({{ $voltageDescr }})'
                     index: 'voltageCurValue.{{ $index }}'
                     low_limit: voltageLowThresh
-                -
-                    oid: transceiverDdmInfoTable
```
Comment (Jellyfrog) [inline includes/definitions/discovery/zynos.yaml:111]: same as above

File: includes/definitions/discovery/zynos.yaml (hunk 1 above)
Comment (Jimmy-Cl) [reply includes/definitions/discovery/zynos.yaml:82]: This discovery has never worked for us, We use two different types of Zynos devices and they use a different set of MIBs. I am happy to leave this in as it should not interfere with my new SFP Dbm discovery.

Comment (Jimmy-Cl) [reply includes/definitions/discovery/zynos.yaml]: Added back so I will resolve this conversation

File: includes/definitions/discovery/zynos.yaml (hunk 2 above)
Comment (Jimmy-Cl) [reply includes/definitions/discovery/zynos.yaml]: Added back so I will resolve this conversation

**Review by murrant — APPROVED** [review]

**Comment by cppmonkey** [comment]
Great to hear the firmware doesn't report CPU load warnings for this.
Had a go back in 2019 but didn't create a PR because of the CPU load issue

https://github.com/cppmonkey/librenms/tree/zynos-transcievers

And Jimmy, check the `ZyXEL LibreNMS Port Mapping` spreadsheet I made. You'll find a much nice way to go from the port ID to switch and port ;) The office just used the first sheet and not the advanced once which contain the following formulas, just in case I didn't leave it behind. 

Switch ID
`=BITRSHIFT(C2,6)+1`

Port ID
`=BITAND(C2,63)+1`

Convert to PHP and we have this to replace Ln13-36 in `includes/discovery/ports/zynos.inc.php`

```
$switchNum = ($portNum >> 6) + 1;
$portNum = ($portNum & 63) + 1;

$ifName = "{$switchNum}/{$portNum}";
```
Can't get much better than bit operations =)

**Comment by cppmonkey** [comment]
Fair play.  Will give it a try and will compare it to my version. See if I can provide some improvements.

Will be possible to gather all the telemetry with the use of PreCache, if that's still a think.  Was going to use it for the Tryin/FS.com EDFA platform... But very difficult with the documentation, MIBs available and that the SNMP responses differed based upon the NMU and/or card firmware revision.
Not used or written and enhancements to LibreNMS in a really long time

**Pull request merged**